pkg/transport/queuestorage/file.go

```diff
@@ -1,138 +1,249 @@
 	"scratchdata/pkg/filestore"
 	"scratchdata/pkg/queue"
 )
 
-type FileEvent struct {
-	Key  string
-	Path string
+const (
+	MaxFileSize int64 = 100 * 1024 * 1024 // 100MB
+	MaxRows     int64 = 1_000
+	MaxFileAge        = 1 * time.Hour
+)
+
+type FileWriterParam struct {
+	Key string
+	Dir string
+
+	MaxFileSize int64
+	MaxRows     int64
+	MaxFileAge  time.Duration
+
+	Queue   queue.QueueBackend
+	Storage filestore.StorageBackend
+}
+
+type FileWriterInfo struct {
+	Key    string
+	Path   string
+	Closed bool
 }
 
 type FileWriter struct {
 	key  string
 	path string
 
-	store  filestore.StorageBackend
-	queue  queue.QueueBackend
-	notify chan FileEvent
-
 	maxFileSize int64
 	maxRows     int64
-	expiry      time.Time
+	maxFileAge  time.Duration
 
-	// Current file being written to
+	queue   queue.QueueBackend
+	storage filestore.StorageBackend
+
+	// fd is the file descriptor of the target file
 	fd *os.File
 
-	// Ensure only 1 file write (or rotate) is happening at a time
-	canWrite sync.Mutex
+	// mu ensure a sequential file change operation
+	mu sync.Mutex
 
-	// Used to rotate every x interval
-	ticker *time.Ticker
+	// terminated is a flag to indicate the file writer is terminated.
+	// When true, all write operations will return a non-nil error
+	// and file rotations will stop.
+	terminated bool
 
-	wg sync.WaitGroup
+	// timer monitors the current file until maxFileAge is reached
+	timer *time.Timer
 }
 
-type NewFileWriterParam struct {
-	Key    string
-	Path   string
-	Store  filestore.StorageBackend
-	Queue  queue.QueueBackend
-	Notify chan FileEvent
-
-	MaxFileSize int64
-	MaxRows     int64
-	MaxFileAge  time.Duration
-}
+func NewFileWriter(param FileWriterParam) (*FileWriter, error) {
+	if param.MaxFileSize == 0 {
+		param.MaxFileSize = MaxFileSize
+	}
+	if param.MaxRows == 0 {
+		param.MaxRows = MaxRows
+	}
+	if param.MaxFileAge == 0 {
+		param.MaxFileAge = MaxFileAge
+	}
 
-func NewFileWriter(param NewFileWriterParam) (*FileWriter, error) {
 	fw := &FileWriter{
 		key:         param.Key,
-		path:        param.Path,
-		store:       param.Store,
-		queue:       param.Queue,
-		notify:      param.Notify,
 		maxFileSize: param.MaxFileSize,
 		maxRows:     param.MaxRows,
+		maxFileAge:  param.MaxFileAge,
 
-		ticker: time.NewTicker(time.Second),
-
-		expiry: time.Now().Add(param.MaxFileAge),
+		queue:   param.Queue,
+		storage: param.Storage,
 	}
 
-	if fw.notify == nil {
-		return nil, fmt.Errorf("notify channel cannot be nil")
-	}
-
-	if err := fw.create(); err != nil {
+	fileName := filepath.Join(param.Dir, fw.key, ulid.Make().String()+".ndjson")
+	if err := fw.create(fileName); err != nil {
 		return nil, err
 	}
 
 	return fw, nil
 }
 
-func (f *FileWriter) create() error {
+// create creates a file and all directories in its path.
+// It sets a timer which triggers close after the maxFileAge elapses.
+func (f *FileWriter) create(fileName string) error {
+	f.path = fileName
 	err := os.MkdirAll(filepath.Dir(f.path), os.ModePerm)
```

Comment:
Not new code, but this uses permission `0777` i.e. read/write/execute for every user on the system. Is this intended? it's typically `0755` or `0700` where potentially sensitive data is being stored.